### DISCUSSION TITLE
[mini.jump] Discussion/suggestion/question

### DISCUSSION THREAD

**Discussion by farzadmf** [nvim-mini/mini.nvim#343] [Ideas]
Hi,

As part of my quest to use as many `mini` modules that I can 😉 , I'm using `mini.jump`.

However I think it somehow "breaks" the habbit of how [neo]vim is supposed to work.

An example:
- Default behavior: when I press `F`, `f`, `T`, `t`, etc. I can go anywhere, press the guy again and repeat the action
- `mini.jump`: if I change my cursor location (even in the same line), I need to tell it _again_ which character I was looking for, which is a bit cumbersome.

Another example:
- Default behavior: if I do a forward/backward search, no matter which one I did, I know that `;` always goes forward and `,` always goes backward.
- `mini.jump`: only allows `;` to repeat, but the issue is that I need to remember which direction I was searching and reverse it if necessary; again, a bit of extra work, in my opinion.

I think my second "issue" can have a workaround by re-pressing the key I used (`f`, `F`, etc.).

So, I was wondering: are these some design decisions that you made to not be compatible with existing behavior?

I think (hence the "idea" category of this topic 😆): it would be nice for the plugin to keep existing built-in behavior, but add on top of that with extra features.

Of course, just a personal opinion and feel free to ignore 🙂

**Comment by echasnovski** [comment]
Hello!

> As part of my quest to use as many `mini` modules that I can wink , I'm using `mini.jump`.

This is a noble quest for sure!

> An example:
> 
>     * Default behavior: when I press `F`, `f`, `T`, `t`, etc. I can go anywhere, press the guy again and repeat the action
> 
>     * `mini.jump`: if I change my cursor location (even in the same line), I need to tell it _again_ which character I was looking for, which is a bit cumbersome.

I am not sure I understand. What you describe as "Default behavior" is not what comes with Neovim as built-in. If you press `fe` it will jump forward on the next `e` inside current line. You can't repeat jump with `f`, it asks again for target. If you want to repeat jump, press `;`.

Yes, target is cleared once cursor moves outside of `FfTt` motions. Mostly because this is the only way to allow user to change target. If it wasn't cleared automatically, `FftT` would only jump on/till one target per session.

> Another example:
> 
>     * Default behavior: if I do a forward/backward search, no matter which one I did, I know that `;` always goes forward and `,` always goes backward.
> 
>     * `mini.jump`: only allows `;` to repeat, but the issue is that I need to remember which direction I was searching and reverse it if necessary; again, a bit of extra work, in my opinion.
> 
> 
> I think my second "issue" can have a workaround by re-pressing the key I used (`f`, `F`, etc.).

Yes, pressing the key you used again or other key to change direction is the design here.

**Reply by farzadmf** [reply]
> What you describe as "Default behavior" is not what comes with Neovim as built-in.

Sorry, I think I wasn't clear there. What I mostly meant is that I can repeat my jump with `;` and `,` with "predictable" results: no matter where I go, I use `;` to jump forward to the same character and `,` to jump backward

Maybe if there was a similar "predictable" behavior in `mini.jump`, it would be nice. As I mentioned, with the way it is, one needs to remember which direction was the last search and `;` repeats that.

Is there any chance to have the forward/backward behavior in `mini.jump`?

**Reply by echasnovski** [reply]
> Sorry, I think I wasn't clear there. What I mostly meant is that I can repeat my jump with `;` and `,` with "predictable" results: no matter where I go, I use `;` to jump forward to the same character and `,` to jump backward

I can not reproduce this with built-in Neovim's `f`, `F`, `t`, and `T`.

After typing `fe`, pressing `;` jumps forward and `,` - backward. Yes.
**But** after typing `Fe`, pressing `;` jumps backward and `,` - forward. This is because `;` is used to "Repeat latest f, t, F or T [count] times." (see `:h ;`) while `,` - same but "... in opposite direction".

I've tried this with `nvim --clean` with the text `  e  e  e`.

Key `;` in 'mini.jump' does the same - repeats in the same direction. But it is mostly there for people who don't want to relearn much from builtins. The designed approach is to repeat the same character (`f`, `F`, `t`, or `T`) to repeat the jump.

**Reply by farzadmf** [reply]
> **But** after typing `Fe`, pressing `;` jumps backward and `,` - forward

Oh wow, you're so right, sorry about that. Time to re-check the docs 😆

Thank you for taking the time for explaining. Keep the modules coming 🙂

**Comment by farzadmf** [comment]
Hey @echasnovski , so I forgot to mention one use case that's possible with built-in behavior, but I don't think is possible in `mini.jump`:
1. In a line, I do `f<char>`
2. Now, `;` will properly remember what I searched for
3. If I want to delete from cursor position to next char, using built-in behavior, I can do `d;`, but that doesn't seem to work in `mini.jump` as it remains in "pending" mode it seems (other combination I use regularly, for example `v;` to visual select seems to work as expected)

**Reply by echasnovski** [reply]
> Hey @echasnovski , so I forgot to mention one use case that's possible with built-in behavior, but I don't think is possible in `mini.jump`:
> 
>     1. In a line, I do `f<char>`
> 
>     2. Now, `;` will properly remember what I searched for
> 
>     3. If I want to delete from cursor position to next char, using built-in behavior, I can do `d;`, but that doesn't seem to work in `mini.jump` as it remains in "pending" mode it seems (other combination I use regularly, for example `v;` to visual select seems to work as expected)

Yes, I can reproduce, thanks! I think the idea was that repeating some edit which uses `;` was not as ergonomic as `df<char>` and using dot-repeat afterwards.

I'll look into it.

**Reply by farzadmf** [reply]
Thank you 🙂

**Reply by echasnovski** [reply]
Using `;` in Operator pending mode (like `d;`, etc.) now shouldn't ask for a target. Thanks for the catch!

**Reply by farzadmf** [reply]
Wow, amazing (as always) and SUPER fast, thank you!
